portal-web/test/functional/com/liferay/portalweb/macros/SearchExperiences.macro

```diff
@@ -0,0 +1,238 @@
+definition {
+
+	macro addBlueprint {
+		Navigator.gotoNavItem(navItem = "Blueprints");
+
+		Click(locator1 = "Button#PLUS");
+
+		PortletEntry.inputTitle(title = "${name}");
+
+		if (isSet(description)) {
+			PortletEntry.inputDescription(description = "${description}");
+		}
+
+		if (isSet(clauseOption)) {
+			LexiconCard.clickCard(card = "Custom Clauses");
+		}
+		else {
+			LexiconCard.clickCard(card = "Liferay Default Clauses");
+		}
+
+		Button.click(button = "Create");
+
+		// Workaround for LPS-133591
+
+		Refresh();
+	}
+
+	macro addElements {
+		Navigator.gotoNavItem(navItem = "Elements");
+
+		Click(locator1 = "Button#PLUS");
+
+		PortletEntry.inputTitle(title = "${name}");
+
+		Button.click(button = "Create");
+
+		// Workaround for LPS-133591
+
+		Refresh();
+
+		SearchExperiences.editElement(name = "${name}");
+
+		Type.typeCodeMirrorEditor(
+			locator1 = "SearchExperiences#BLUEPRINTS_ELEMENT_TEMPLATE_JSON",
+			value1 = "${elementTemplateJSON}");
+
+		Type.typeCodeMirrorEditor(
+			locator1 = "SearchExperiences#BLUEPRINTS_ELEMENT_UI_CONFIGURATION_JSON",
+			value1 = "${uiConfigurationJSON}");
+
+		PortletEntry.save();
+	}
+
+	macro addQueryElements {
+		AssertElementPresent(locator1 = "//div[contains(@class,'sidebar-header')]//span[contains(.,'Add Query Elements')]");
+
+		Type.sendKeysApplicationSearch(
+			locator1 = "SearchExperiences#BLUEPRINTS_FRAGMENT_QUERY_SEARCH_BAR",
+			value1 = "${fragment}");
+
+		MouseOver(
+			key_fragmentName = "${fragment}",
+			locator1 = "SearchExperiences#BLUEPRINTS_QUERY_FRAGMENT_NAME");
+
+		Click.clickNoMouseOverNoWaitForVisible(
+			key_fragmentName = "${fragment}",
+			locator1 = "SearchExperiences#BLUEPRINTS_ADD_QUERY_FRAGMENT");
+
+		if (isSet(assetType)) {
```

Comment:
I think adding the configurations to the testcase would avoid making a monster macro but I also think it would make the testcase look confusing especially since there are different kind of options to add, dropdown, and type. Do you think we can add a comment or something at the top of each addition explaining the element field or is that bad practice?

Reply:
hm yeah I can see how it would make testcases confusing for elements with heavy configuration. not sure what we should do with this macro now though. adding comments might help a bit, but that doesn't alleviate the difficulty of maintaining and scaling the macro. one example I can think of is you wouldn't be able to add and configure two of the same element. maybe we can look into how other teams handle this for similar features they have. but for now, i'll just leave the conditionals alone.

Reply:
we could always extract each configuration into separate macros and call them when needed. we may end up with one liner macros but it breaks down the "one macro to rule them all" into managable chunks and makes things on the testcase level more readable